dev/src/document-reader.ts

```diff
@@ -0,0 +1,193 @@
+ * failure and returns ordered results.
+ *
+ * @private
+ */
+export class DocumentReader<T> {
+  /** An optional field mask to apply to this read. */
+  fieldMask?: FieldPath[];
+  /** An optional transaction ID to use for this read. */
+  transactionId?: Uint8Array;
+
+  private remainingDocuments = new Set<string>();
+  private retrievedDocuments = new Map<string, DocumentSnapshot>();
+
+  /**
+   * Internal method to retrieve multiple documents from Firestore, optionally
+   * as part of a transaction.
+   *
+   * @param firestore The Firestore instance to use.
+   * @param allDocuments The documents to receive.
+   * @returns A Promise that contains an array with the resulting documents.
+   */
+  constructor(
+    private firestore: Firestore,
+    private allDocuments: Array<DocumentReference<T>>
+  ) {
+    for (const docRef of this.allDocuments) {
+      this.remainingDocuments.add(docRef.formattedName);
+    }
+  }
+
+  /**
+   * Invokes the BatchGetDocuments RPC and returns the results.
+   *
+   * @param requestTag A unique client-assigned identifier for this request.
+   */
+  async get(requestTag: string): Promise<Array<DocumentSnapshot<T>>> {
+    await this.fetchAllDocuments(requestTag);
+
+    // BatchGetDocuments doesn't preserve document order. We use the request
+    // order to sort the resulting documents.
+    const orderedDocuments: Array<DocumentSnapshot<T>> = [];
+
+    for (const docRef of this.allDocuments) {
+      const document = this.retrievedDocuments.get(docRef.formattedName);
+      if (document !== undefined) {
+        // Recreate the DocumentSnapshot with the DocumentReference
+        // containing the original converter.
+        const finalDoc = new DocumentSnapshotBuilder(
+          docRef as DocumentReference<T>
+        );
+        finalDoc.fieldsProto = document._fieldsProto;
+        finalDoc.readTime = document.readTime;
+        finalDoc.createTime = document.createTime;
+        finalDoc.updateTime = document.updateTime;
+        orderedDocuments.push(finalDoc.build());
+      } else {
+        throw new Error(`Did not receive document for "${docRef.path}".`);
+      }
+    }
+
+    return orderedDocuments;
+  }
+
+  private async fetchAllDocuments(requestTag: string): Promise<void> {
+    while (this.remainingDocuments.size > 0) {
+      try {
+        return await this.fetchMoreDocuments(requestTag);
+      } catch (err) {
+        // If a non-transactional read failed, attempt to restart.
+        // Transactional reads are retried via the transaction runner.
+        if (
+          this.transactionId ||
+          isPermanentRpcError(err, 'batchGetDocuments')
+        ) {
+          logger(
+            'DocumentReader.fetchAllDocuments',
+            requestTag,
+            'BatchGetDocuments failed with non-retryable stream error:',
+            err
+          );
+          throw err;
+        } else {
+          logger(
```

Comment:
Is the maximum number of retries allowed capped?

Reply:
No, I thought it would be better not to. But then I changed my mind. I rewrote this logic to only fetch more results if we made some progress, which should ensure that we never end up in a loop with just errors. This also removes the need to add backoff, which I otherwise should have done.

FYI - This rewrite was pretty complicated, so I ended up changing the implementation to use async iterators. I did this in steps in case you want to see more reasonable diffs.